keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,597 @@
+1. `FeatureA` is set to `val` (`val=true|false`) in an existing cluster.
+1. A new member of newer minor or batch version joins with the code path `FeatureA=val` removed.
+1. In this case, the new member would not be able to handle the existing data it receives requiring the code path `FeatureA=val`.
+
+To avoid this problem, we need to ensure that in the previous version, `FeatureA` can only be set to the fixed value `!val`, because etcd supports 1 minor version skew in a cluster.
+
+So to make it possible to remove a feature, we require a feature to go through the following 3 stages toward deprecation before it can be removed:
+1. Feature has been stable in the previous lifecycle for at least 1 minor release (no change of default values in any patch version of this minor version). This is to ensure we would not have to change its default value in the next stage.
+1. Feature has been stable in lifecycle `PreRelease: Deprecated, {Default: defaultVal, LockToDefault: true` for at least 1 minor release.
+1. Feature can be removed in the next release.
+
+The figure below shows an example of how a GA feature can be removed.
+![ga feature removal](./etcd_ga_feature_removal.png "ga feature removal")
+
+One corner case we need to consider is: when a new cluster starts, before it establishes the final `ClusterVersion`, the `ClusterVersion` will be set nil first then `MinClusterVersion=3.0`. For both cases `ClusterParams` would be nil and no feature would be enabled. If the code path for `FeatureA=false` is removed in some servers, and still exists in some other servers, we need to make sure `FeatureA=true` even when `ClusterVersion == nil|"3.0"`. 
+There are two ways we can solve this issue: 
+1. the leader would need to use the default value of the feature if `PreRelease == Deprecated && LockToDefault == true` at the binary version even when `ClusterVersion == nil|MinClusterVersion`. This assumes a feature would not be removed unless it has been stable at `PreRelease == Deprecated && LockToDefault == true` lifecycle for at least 1 minor version.
+    ```go
+    func (c *RaftCluster) FeatureEnabled(f featuregate.Feature) {
+      if c.Params != nil && enabled, ok := c.Params.FeatureGates[f]; ok {
+        return enabled
+      }
+      if c.version == nil || c.version == MinClusterVersion {
+        featureSpec := clusterFeatureGate.GetAll(binaryVersion)[f]
+        if featureSpec.PreRelease == Deprecated && featureSpec.LockToDefault == true {
+          return featureSpec.Default
+        }
+      }
+      return false
+    }
+    ```
+2. make the cluster not ready to serve until the actual `ClusterCluster` is established. This would ensure there is no version jump from `3.0` to `3.6` for example, and the cluster is not serving traffic when `ClusterParams == nil`. But this would break the HA assumption (meaning the cluster should be able to server when the majority of servers are online) during bootstrapping.
```

Comment:
Would like to revisit this, is there any reason we can't set clusterVersion based on quorum of available clusters instead of all needing to be up? It creates a risk that cluster version we be set higher before we have older members join, however I don't know why we would support hybrid version cluster bootstrap.